### DISCUSSION TITLE
[Chapter6] 입출력이 없는 컴퓨터가 있을까 ?

### DISCUSSION THREAD

**Discussion by yunseorim1116** [elegant-functional-2023/secrets-of-computer#15] [General]
감상평을 남겨주세요~
범위: 섹션 6.1-6.7 (끝까지)
8/10 15시 전까지~! ^ㅇ^

**Comment by thgr8ganzi** [comment]
## 감상평

* 이번편은 좀 편하게 읽은것 같아요
* 벌써 책이 끝났다니 기쁘네요
* 디스크가 DMA 를 통해서 알아서 데이터를 전송한다는게 신기하네요
* 나는 코드만 짜면 컴퓨퓨터가 이 모든걸 알아서 해준다니 컴퓨터야 고마워

**Reply by iborymagic** [reply]
퓨퓨퓨

**Reply by hanana1253** [reply]
![image](https://github.com/user-attachments/assets/592c47ea-e70a-4dde-a327-f3499ee4c838)

**Comment by yunseorim1116** [comment]
잊어버릴까봐 후다닥 남기고 나중에 수정하겠습니다,,,,,,,
이번 파트는 집중을 못해서 다시 한 번 읽을 예정입니다
마법처럼 느껴졌던 컴푸타..컴퓨터 동작에 대해 발가락을 담가본거같아 기분이 좋았습니다 
[정리링크](https://nifty-period-b9a.notion.site/9-6a834441bdea45d5852f1038791d0cef?pvs=25)

**Reply by sunhpark42** [reply]
제가 서림님 링크 감사히 사용해서 6.1 리딩 하겠습니다.

**Reply by sunhpark42** [reply]
🦶

**Reply by thgr8ganzi** [reply]
빨리 나으세요

**Reply by ddongule** [reply]
6.2 감사합니다

**Comment by ella-yschoi** [comment]
## 감상평

- 책에서 중간중간 질문을 던져주며 개념을 확장해가서 좋았습니다.

<br/>

## 6.1장: CPU는 어떻게 입출력 작업을 처리할까?

- CPU가 어떻게 장치의 작업 상태를 알 수 있는지 두 가지 방법 (폴링, 인터럽트)을 소개하고 있군요.
- 인터럽트는 앞에서도 많이 다룬 내용이라 비교적 어렵지 않게 읽었습니다.
- 중단된 프로그램의 실행 상태와 복원 시, 스택 개념이 적용된다는게 흥미로웠습니다. 지금보니 당연한 것 같기도?


<br/>

## 6.2장: 디스크가 입출력을 처리할 때 CPU가 하는 일은 무엇일까?

- 역시 결론은 비동기가 최고군요
- DMA는 원래 CPU가 해야 할 작업의 일부를 Dㅐ신한다
- 변수 간의 일관성 문제는 앞에서도 언급되었던 거라 반가웠습니다.

<br/>

## 6.3장: 파일을 읽을 때 프로그램에는 어떤 일이 발생할까?

- 요약: 메모리와 외부 장치 사이에 복사 데이터가 왔다갔다 하는 것을 입출력 (I/O) 이라고 한다.
- 운영체제 덕분에 모두가 바쁘고 효율적으로 움직일 수 있다는 것을 알려준 챕터였습니다.

<br/>

## 6.4장: 높은 동시성의 비결: 입출력 다중화

- 파일 서술자에 대해 이야기할 때 대기번호 비유가 잘 와닿았던 것 같아요.
- 다중스레드 생각을 하긴 했는데 책에서 말한 것처럼 스케줄링 시 부담과 동시성이 필요한 상황에는 최적의 방법이 아니긴 하네요.
- select, poll, epoll 을 비교해서 설명을 들으니 더 이해가 잘 됐습니다. 

<br/>

## 6.5장: mmap: 메모리 읽기와 쓰기 방식으로 파일 처리하기

- 평소 터미널에 ls 명령어를 입력할 때 일어나는 일을 알 수 있었습니다. 하지만 이해하기 어렵기도 했어요.

<br/>

## 6.6장: 컴퓨터 시스템의 각 부분에서 얼마큼 지연이 일어날까?

- 약간.. '이런 것까지 알아야 할까?' 싶었던 내용.
어쨌든 말하고자 하는건 컴퓨터 시스템의 주요 작업에서 발생하는 시간은 매년 줄어들고 있다, 메모리에 접근하는 속도가 SSD에 접근하는 속도보다 빠르다, CPU 속도에 비해 메모리 접근이 매우 느리다 라는 말을 수치적으로 하는 것 같아요.

<br/>

**Reply by hanana1253** [reply]
Dㅐ신한다 ㅋㅋㅋ

**Reply by thgr8ganzi** [reply]
Dㅐ신한다

**Comment by ddongule** [comment]
# 와 책 끝 났 다!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
- 저두 이번 편은 좀 뭐랄까 머리를 쓴다기보다능 약간 술술? 읽힌거같아요 (뭔가 나와 멀리 떨어져있는 이야기 같아서 ^^ . . . 그런듯 ㅋ ^^ . . . )
- 항상 컴퓨터를 보면서 마법같았는데 그 마법의 실마리가 풀리지는 않았지만 어쨋든 이것도 사람이 만든 엇쩌구다ㅋ라는걸 알게되엇다ㅋ 
- 책 다시 읽엉야겠따. 그땐 더 이해를 잘 할 수 있겠지?
- 모든 컴퓨터공학과 교수님들이 이걸로 강의해준다면 우리나라에도 빌게이츠가 몇명 생기지 않을까?
- 사상 얘기가 너무 많이 나와서 이제 사상-mapping 이라고 잘 읽힌다. 첨에 나왓을때 진짜 안읽혔는데 우헤헥
- 6.6에서 시간 지표로 환산 까지는 머. . . 애쓴다 싶었는데 거리 지표로 환산한거 보고 뇌절이다 싶엇다 ㅋ
- 끝 났 따!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! 오예 회식~~~~~~~~~~

**Reply by sunhpark42** [reply]
‼️‼️‼️‼️‼️‼️‼️‼️‼️‼️‼️‼️‼️‼️‼️‼️‼️‼️‼️‼️‼️‼️‼️‼️‼️‼️‼️‼️

**Reply by iborymagic** [reply]
글 사운드 좀 줄여주세요

**Reply by sunhpark42** [reply]
ㅋㅋㅋㅋㅋ민경게이츠

**Reply by hanana1253** [reply]
앞으로 일할 때 "사상해주세요"

**Comment by hanana1253** [comment]
마지막장이라 신났고 비교적 이해하기 어렵지 않았다. 스터디가 오후에 시작이니까 충분히 읽고 쉬고 올 수 있을거라 자신했지만... 항상 스터디 2.5시간 전 시작하는 것이 몸에 프로그래밍 되어있다는 것을 깨달았다. 

계속해서 CPU를 허드렛일에 쓰지 않는단 걸 강조하는 걸 보고 CPU는 함부로 갖다쓰는게 아닌 고오급인력이라는 생각이 들었다. 나도 그런 인력이 되고 싶다. 그치만 CPU처럼 쉼없이 일하게 되는 건 또 싫다. 

열 놈에 죽 한 사발이라는 속담 웃기다. 우린 오늘 한 놈에 죽 열사발처럼 먹어버리는 회식을 할 것이다.

**Reply by ddongule** [reply]
죽? no . 마늘 basasak yes.

**Reply by sunhpark42** [reply]
ㅋㅋㅋㅋㅋㅋㅋㅋㅋ회식..기대됩니다

**Reply by ella-yschoi** [reply]
정말 한 놈에 죽 열사발처럼 먹어버리는 회식을 해버렸네요 ㅋㅋㅋㅋㅋㅋㅋㅋ

**Comment by chloe-codes1** [comment]
## 감상평

- I/O 가 일어날 때의 동작 방식에서 인터럽트 거는 것 정도로 간단히 알았는데 상세히 알 수 있어 좋았다
- 추가로 찾아본 것
  - 파일 데이터가 각 프로세스 주소공간에 복사될때 운영체제 내부의 page cache 와 buffer cache 로 저장해서 I/O 성능을 최적화 하는 부분! 
  - File descriptor 는 Kernel에서 생성되고, file descriptor table 에서의 index 역할을 한다
  - File descriptor 는 연속적으로 증가하는 정수 번호로 할당되고, 파일이 닫히면 해당 번호는 재사용 가능하다


끝났다~~~~~! 뿌듯 ⁺◟( ᵒ̴̶̷̥ ·̫ ᵒ̴̶̷̣ )

**Reply by sunhpark42** [reply]
⁺◟( ᵒ̴̶̷̥ ·̫ ᵒ̴̶̷̣ )

**Comment by iborymagic** [comment]
- 코딩할 때는 인지를 못하고 있었는데, 파일 입출력 과정이 생각보다 되게 복잡하구나 하는 것을 알게 되었읍니다..
- 정말 프로세스, 스레드, CPU 등등 하드웨어와 소프트웨어가 굉장히 촘촘히 엮여서 동작하고 있다는 것을 다시 한 번 깨닫게 되었읍니다.. 마지막까지 읽다보니 경이롭다는 생각이 들면서도 그걸 어렵지 않게(쉽게는 아니고) 설명해주는 글쓴이도 참 대단하다고 생각이 들었읍니다..
- 근데 DMA는 작동 방식이라고 했는데 하드웨어 장치처럼 설명되는 것 같더라구요? 동작 방식과 하드웨어 장치의 이름이 동일한 것이겠지요?
- 이 책을 꿰뚫는 핵심 키워드는 어쩌면 가상 메모리일지도..
- `모든 것은 파일이다` 라는 것도 뭔가 인상깊었지만 제대로 이해는 못했읍니다.. 다음에 더 자세히 찾아봐야 할까봐요..

**Comment by bigsaigon333** [comment]
- 단순(?) 키보드 입력에도 인터럽트 라는 cpu 메커니즘이 있다
- 마지막장 너무 집중이 안돼요 😭 감사합니다 브라우저

**Comment by sunhpark42** [comment]
* CPU 일.. 고생하고있구나 짜식...
* 운영체제 네트워크 뭔가 더 있는데 휘리릭 한 것 같은 기분이 드는 6장이었슴돠
* 입출력 인터럽트 설명하던 교수님이 조금 생각납니다.. 이게 다 어제 같은학교 컴퓨터학과 사람들이랑 술마셔서 그럼

**Reply by sunhpark42** [reply]
https://www.youtube.com/@user-mi9bu4lm6g/videos 교수님 강의를 볼 수 있다 근데 2019년도임

**Comment by sa02045** [comment]
- CPU를 최대한으로 사용하기 위한 방법을 또 배울 수 있었음....(I/O가 발생해도 CPU를 열심히 일시키는 방법.. CPU야 열심히 일해줘!)
- node.js 공부할 때 나왔던 I/O 논블락킹이라는 단어를 좀 더 이해할 수 있었음
- 모든 것은 파일이다라는 것은 알고 있었는데, 입출력도 파일로 다룬다는 것이 재밌었음
- 동기적인 문제를 비동기적으로 해결한다는 아이디어는 저수준계층이나 고수준계층이나 모두 통하는 해결방법인 것 같음
- 책에 제프 딘 아저씨 나와서 반가웠음!
